pyfarm/master/user_interface/software.py

```diff
@@ -0,0 +1,191 @@
+        error = ("This software cannot be deleted, it is still in use by the "
+                 "following jobs or jobtypes: ")
+        first = True
+        for job_requirement in job_requirements:
+            if not first:
+                error.append(", ")
+            first = False
+            error.append("job %s" % job_requirement.job.title)
+        for jobtype_requirements in jobtype_requirements:
+            if not first:
+                error.append(", ")
+            first = False
+            error.append("jobtype %s" %
+                         jobtype_requirements.jobtype_version.jobtype.name)
+        return render_template("pyfarm/error.html", error=error), BAD_REQUEST
+
+    db.session.delete(software)
+    db.session.commit()
+
+    flash("Software %s has been removed" % software.software)
+
+    return redirect(url_for("software_index_ui"), SEE_OTHER)
+
+def update_version_rank(software_id, version_id):
+    software = Software.query.filter_by(id=software_id).first()
+    if not software:
+        return (render_template("pyfarm/error.html",
+                                error="Software %s not found" % software_id),
+                NOT_FOUND)
+
+    version = SoftwareVersion.query.filter_by(
+        software=software, id=version_id).first()
+    if not version:
+        return (render_template("pyfarm/error.html",
+                                error="Software %s version %s not found" %
+                                    (software_id, version_id)),
+                NOT_FOUND)
+
+    version.rank = int(request.form["rank"])
+
+    db.session.add(version)
+    db.session.commit()
+
+    flash("Rank for verson %s has been updated." % version.version)
+
+    return redirect(url_for("single_software_ui", software_id=software.id),
+                    SEE_OTHER)
+
+def remove_software_version(software_id, version_id):
+    software = Software.query.filter_by(id=software_id).first()
+    if not software:
+        return (render_template("pyfarm/error.html",
+                                error="Software %s not found" % software_id),
+                NOT_FOUND)
+
+    version = SoftwareVersion.query.filter_by(
+        software=software, id=version_id).first()
+    if not version:
+        return redirect(url_for("single_software_ui", software_id=software.id),
+                    SEE_OTHER)
+
+    job_requirements = JobSoftwareRequirement.query.filter(
+        or_(JobSoftwareRequirement.min_version == version,
+            JobSoftwareRequirement.max_version == version)).all()
+    jobtype_requirements = JobTypeSoftwareRequirement.query.filter(
+        or_(JobTypeSoftwareRequirement.min_version == version,
+            JobTypeSoftwareRequirement.max_version == version)).all()
+    if job_requirements or jobtype_requirements:
+        error = ("This version cannot be deleted, it is still in use by the "
+                 "following jobs or jobtypes: ")
+        first = True
```

Comment:
You could probably simplify this logic by appending to a list instead of the original error string.  Then you'd just do `", ".join(errors)`